presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -211,17 +216,68 @@ public PlanNode visitProject(ProjectNode node, RewriteContext<Expression> contex
+            // certain limitations.
+            List<Expression> deterministicConjuncts = conjuncts.get(true);
 
-            // All non-deterministic conjuncts, if any, will be in the filter node.
+            // We partition the expressions in the deterministicConjuncts into two lists, and only inline the
+            // expressions that are in the inlining targets list.
+            Map<Boolean, List<Expression>> partitionedConjuncts = deterministicConjuncts.stream()
+                    .collect(Collectors.partitioningBy(expression -> isInliningCandidate(expression, node)));
+
+            List<Expression> inlinedDeterministicConjuncts = partitionedConjuncts.get(true).stream()
+                    .map(entry -> inlineSymbols(node.getAssignments().getMap(), entry))
+                    .collect(Collectors.toList());
+
+            PlanNode rewrittenNode = context.defaultRewrite(node, combineConjuncts(inlinedDeterministicConjuncts));
+
+            // All deterministic conjuncts that contains non-inlining targets, and non-deterministic conjuncts,
+            // if any, will be in the filter node.
+            List<Expression> nonInliningConjuncts = partitionedConjuncts.get(false);
             if (!conjuncts.get(false).isEmpty()) {
-                rewrittenNode = new FilterNode(idAllocator.getNextId(), rewrittenNode, combineConjuncts(conjuncts.get(false)));
+                nonInliningConjuncts.addAll(conjuncts.get(false));
+            }
+
+            if (!nonInliningConjuncts.isEmpty()) {
+                rewrittenNode = new FilterNode(idAllocator.getNextId(), rewrittenNode, combineConjuncts(nonInliningConjuncts));
             }
 
             return rewrittenNode;
         }
 
+        private boolean isInliningCandidate(Expression expression, ProjectNode node)
+        {
+            Set<Symbol> childOutputSet = ImmutableSet.copyOf(node.getOutputSymbols());
+
+            // candidate symbols for inlining are
+            //   1. references to simple constants
+            //   2. references to complex expressions that
+            //      a. are not inputs to try() expressions
+            //      b. appear only once across all expressions
+            // which come from the node, as opposed to an enclosing scope.
+
+            Map<Symbol, Long> dependencies = SymbolsExtractor.extractAll(expression).stream()
+                    .filter(childOutputSet::contains)
+                    .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+
+            // exclude any complex inputs to TRY expressions. Inlining them would potentially
+            // change the semantics of those expressions
+            Set<Symbol> tryArguments = extractTryArguments(expression);
+
+            return dependencies.entrySet().stream()
+                    .allMatch(entry -> (entry.getValue() == 1 || node.getAssignments().get(entry.getKey()) instanceof Literal)
```

Comment:
is not `1` too restrictive? Mentioned issue was about inlining when number of occurrences grew exponentially. Maybe we could have 10 here? Inlining expression 10 times should not affect planning but it should allow  to handle more predicate pushdowns.

Reply:
No, you get exponential behavior with anything > 1 depending on the depth of the plan. In the example shown in  #10455 the number of occurrences is 3 for each expression.

Reply:
True, thanks for explanation.